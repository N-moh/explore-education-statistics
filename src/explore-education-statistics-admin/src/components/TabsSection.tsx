import classNames from 'classnames';
import React, {
  forwardRef,
  FunctionComponent,
  HTMLAttributes,
  ReactNode,
  Ref,
  useState,
} from 'react';
import useRendered from '../hooks/useRendered';
import styles from './TabsSection.module.scss';

export interface TabsSectionProps {
  children: ReactNode;
  id: string;
  /**
   * Set to true if children should not be
   * rendered until tab has been selected.
   */
  lazy?: boolean;
  title: string;
}

const TabsSection: FunctionComponent<TabsSectionProps> = forwardRef(
  (
    { children, id, lazy = false, ...restProps }: TabsSectionProps,
    ref: Ref<HTMLElement>,
  ) => {
    const [mouseDown, setMouseDown] = useState(false);
    const { onRendered } = useRendered();

    // Hide additional props from the component's public API to
    // avoid any confusion over this component's usage as
    // it should only be used in combination with Tabs.
    const tabProps = restProps as HTMLAttributes<HTMLElement>;

    return (
      <section
        aria-labelledby={onRendered(tabProps['aria-labelledby'])}
        className={classNames('govuk-tabs__panel', styles.panel, {
          'govuk-tabs__panel--hidden': tabProps.hidden,
        })}
        id={id}
        hidden={tabProps.hidden}
        ref={ref}
        onFocus={event => {
          if (mouseDown) {
            event.target.blur();
          }
        }}
        onMouseDown={() => setMouseDown(true)}
        onMouseUp={() => setMouseDown(false)}
        role={onRendered('tabpanel')}
        tabIndex={onRendered(-1)}
      >
        {children}
      </section>
    );
  },
);

TabsSection.displayName = 'TabsSection';

export default TabsSection;
