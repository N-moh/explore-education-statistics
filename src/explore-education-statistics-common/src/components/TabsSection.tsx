import { useDesktopMedia } from '@common/hooks/useMedia';
import classNames from 'classnames';
import React, {
  createElement,
  forwardRef,
  HTMLAttributes,
  ReactNode,
} from 'react';
import styles from './TabsSection.module.scss';

export interface TabsSectionProps {
  children: ReactNode;
  id: string;
  /**
   * Set to true if children should not be
   * rendered until tab has been selected.
   */
  lazy?: boolean;
  title: string;
  headingTag?: 'h2' | 'h3' | 'h4';
}

const TabsSection = forwardRef<HTMLElement, TabsSectionProps>(
  (
    { children, id, title, headingTag = 'h3', ...restProps }: TabsSectionProps,
    ref,
  ) => {
    const { onMedia } = useDesktopMedia();

    // Hide additional props from the component's public API to
    // avoid any confusion over this component's usage as
    // it should only be used in combination with Tabs.
    const tabProps = restProps as HTMLAttributes<HTMLElement>;

    return (
      <section
        aria-labelledby={onMedia(tabProps['aria-labelledby'])}
        className={classNames(
          'govuk-tabs__panel',
          'dfe-content-overflow',
          styles.panel,
          {
            'govuk-tabs__panel--hidden': tabProps.hidden,
          },
        )}
        id={id}
        ref={ref}
        role={onMedia('tabpanel')}
        tabIndex={onMedia(-1)}
        data-testid={title}
      >
        {createElement(headingTag, { children: title })}
        {children}
      </section>
    );
  },
);

TabsSection.displayName = 'TabsSection';

export default TabsSection;
