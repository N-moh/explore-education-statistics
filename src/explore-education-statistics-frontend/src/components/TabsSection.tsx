import React, { FunctionComponent } from 'react';

export interface TabsSectionProps {
  id: string;
  title: string;
}

const TabsSection: FunctionComponent<TabsSectionProps> = ({
  children,
  title,
  id,
}) => {
  return (
    <div key={id} className="govuk-tabs__panel" id={id}>
      {children}
    </div>
  );
};

export default TabsSection;
