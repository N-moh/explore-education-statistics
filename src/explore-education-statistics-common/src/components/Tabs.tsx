import { useDesktopMedia } from '@common/hooks/useMedia';
import isComponentType from '@common/lib/type-guards/components/isComponentType';
import classNames from 'classnames';
import React, {
  cloneElement,
  HTMLAttributes,
  ReactElement,
  ReactNode,
  RefAttributes,
  useState,
} from 'react';
import TabsSection, { TabsSectionProps } from './TabsSection';

function filterSections(children: ReactNode): ReactElement<TabsSectionProps>[] {
  if (Array.isArray(children)) {
    return children.filter(child =>
      isComponentType(child, TabsSection),
    ) as ReactElement<TabsSectionProps>[];
  }

  if (isComponentType(children, TabsSection)) {
    return [children];
  }

  return [];
}

interface Props {
  children: ReactNode;
}

const Tabs = ({ children }: Props) => {
  const [loadedSections, setLoadedSections] = useState(new Set<number>());
  const [selectedTabIndex, setSelectedTabIndex] = useState(0);

  const { onMedia } = useDesktopMedia();

  const tabElements: HTMLAnchorElement[] = [];
  const sectionElements: HTMLElement[] = [];

  const setSelectedTab = (index: number) => {
    window.location.hash = sectionElements[index].id;

    setLoadedSections(loadedSections.add(index));
    setSelectedTabIndex(index);
  };

  const sections = filterSections(children);

  return (
    <div className="govuk-tabs">
      <ul className="govuk-tabs__list" role="tablist">
        {sections.map(({ props }, index) => (
          <li
            className="govuk-tabs__list-item"
            key={props.id}
            role="presentation"
          >
            <a
              aria-controls={onMedia(props.id)}
              aria-selected={onMedia(selectedTabIndex === index)}
              role={onMedia('tab')}
              className={classNames('govuk-tabs__tab', {
                'govuk-tabs__tab--selected': selectedTabIndex === index,
              })}
              href={`#${props.id}`}
              id={`${props.id}-tab`}
              ref={(element: HTMLAnchorElement) => tabElements.push(element)}
              tabIndex={selectedTabIndex !== index ? -1 : undefined}
              onClick={event => {
                event.preventDefault();
                setSelectedTab(index);
              }}
              onKeyDown={event => {
                switch (event.key) {
                  case 'ArrowLeft': {
                    const nextTabIndex =
                      selectedTabIndex > 0
                        ? selectedTabIndex - 1
                        : sections.length - 1;

                    setSelectedTab(nextTabIndex);
                    tabElements[nextTabIndex].focus();

                    break;
                  }
                  case 'ArrowRight': {
                    const nextTabIndex =
                      selectedTabIndex < sections.length - 1
                        ? selectedTabIndex + 1
                        : 0;

                    setSelectedTab(nextTabIndex);
                    tabElements[nextTabIndex].focus();

                    break;
                  }
                  case 'ArrowDown':
                    sectionElements[selectedTabIndex].focus();
                    break;
                  default:
                    break;
                }
              }}
            >
              {props.title}
            </a>
          </li>
        ))}
      </ul>

      {sections.map((section, index) => {
        const { id, lazy } = section.props;

        return cloneElement<
          TabsSectionProps &
            HTMLAttributes<HTMLElement> &
            RefAttributes<HTMLElement>
        >(
          section,
          {
            'aria-labelledby': `${id}-tab`,
            hidden: selectedTabIndex !== index,
            key: id,
            ref: (element: HTMLElement) => sectionElements.push(element),
          },
          lazy && !loadedSections.has(index) ? null : section.props.children,
        );
      })}
    </div>
  );
};

export default Tabs;
