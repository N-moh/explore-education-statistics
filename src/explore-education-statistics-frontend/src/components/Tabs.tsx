import classNames from 'classnames';
import React, {
  cloneElement,
  Component,
  HTMLAttributes,
  ReactElement,
  ReactNode,
  RefAttributes,
} from 'react';
import isComponentType from '../lib/type-guards/components/isComponentType';
import TabsSection, { TabsSectionProps } from './TabsSection';

interface Props {
  children: ReactNode;
}

interface State {
  loadedSections: Set<number>;
  selectedTabIndex: number;
}

class Tabs extends Component<Props, State> {
  public state = {
    loadedSections: new Set<number>(),
    selectedTabIndex: 0,
  };

  private tabElements: HTMLAnchorElement[] = [];
  private sectionElements: HTMLElement[] = [];

  private setSelectedTab(index: number) {
    this.setState({
      loadedSections: this.state.loadedSections.add(index),
      selectedTabIndex: index,
    });
  }

  public render() {
    const { children } = this.props;
    const { selectedTabIndex } = this.state;

    const sections = this.filterSections(children);

    return (
      <div className="govuk-tabs">
        <ul className="govuk-tabs__list" role="tablist">
          {sections.map(({ props }, index) => (
            <li
              className="govuk-tabs__list-item"
              key={props.id}
              role="presentation"
            >
              <a
                aria-controls={props.id}
                aria-selected={selectedTabIndex === index}
                className={classNames('govuk-tabs__tab', {
                  'govuk-tabs__tab--selected': selectedTabIndex === index,
                })}
                href={`#${props.id}`}
                id={`${props.id}-tab`}
                ref={(element: HTMLAnchorElement) =>
                  this.tabElements.push(element)
                }
                role="tab"
                onClick={event => {
                  event.preventDefault();
                  this.setSelectedTab(index);
                }}
                onKeyDown={event => {
                  switch (event.key) {
                    case 'ArrowLeft': {
                      const nextTabIndex =
                        selectedTabIndex > 0
                          ? selectedTabIndex - 1
                          : sections.length - 1;

                      this.setSelectedTab(nextTabIndex);
                      this.tabElements[nextTabIndex].focus();

                      break;
                    }
                    case 'ArrowRight': {
                      const nextTabIndex =
                        selectedTabIndex < sections.length - 1
                          ? selectedTabIndex + 1
                          : 0;

                      this.setSelectedTab(nextTabIndex);
                      this.tabElements[nextTabIndex].focus();

                      break;
                    }
                    case 'ArrowDown':
                      this.sectionElements[selectedTabIndex].focus();
                      break;
                  }
                }}
                tabIndex={selectedTabIndex !== index ? -1 : undefined}
              >
                {props.title}
              </a>
            </li>
          ))}
        </ul>

        {sections.map((section, index) => {
          const { id, lazy } = section.props;

          return cloneElement<
            TabsSectionProps &
              HTMLAttributes<HTMLElement> &
              RefAttributes<HTMLElement>
          >(
            section,
            {
              'aria-labelledby': `${id}-tab`,
              hidden: selectedTabIndex !== index,
              key: id,
              ref: (element: HTMLElement) => this.sectionElements.push(element),
            },
            lazy && !this.state.loadedSections.has(index)
              ? null
              : section.props.children,
          );
        })}
      </div>
    );
  }

  private filterSections(
    children: ReactNode,
  ): ReactElement<TabsSectionProps>[] {
    if (Array.isArray(children)) {
      return children.filter(child =>
        isComponentType(child, TabsSection),
      ) as ReactElement<TabsSectionProps>[];
    }

    if (isComponentType(children, TabsSection)) {
      return [children];
    }

    return [];
  }
}

export default Tabs;
