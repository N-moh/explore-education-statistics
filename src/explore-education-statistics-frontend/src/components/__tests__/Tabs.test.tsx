import React from 'react';
import { fireEvent, render } from 'react-testing-library';
import Tabs from '../Tabs';
import TabsSection from '../TabsSection';

describe('Tabs', () => {
  describe('multiple tabs', () => {
    test('renders correctly', () => {
      const { container } = render(
        <Tabs>
          <TabsSection id="tab-1" title="Tab 1">
            <p>Test section 1 content</p>
          </TabsSection>
          <TabsSection id="tab-2" title="Tab 2">
            <p>Test section 2 content</p>
          </TabsSection>
        </Tabs>,
      );

      expect(container.innerHTML).toMatchSnapshot();
    });

    test('tab links match section ids', () => {
      const { getByText } = render(
        <Tabs>
          <TabsSection id="tab-1" title="Tab 1">
            <p>Test section 1 content</p>
          </TabsSection>
          <TabsSection id="tab-2" title="Tab 2">
            <p>Test section 2 content</p>
          </TabsSection>
        </Tabs>,
      );

      expect(getByText('Tab 1')).toHaveAttribute('href', '#tab-1');
      expect(getByText('Tab 2')).toHaveAttribute('href', '#tab-2');
    });

    test('clicking tab reveals correct section', () => {
      const { getByText, container } = render(
        <Tabs>
          <TabsSection id="tab-1" title="Tab 1">
            <p>Test section 1 content</p>
          </TabsSection>
          <TabsSection id="tab-2" title="Tab 2">
            <p>Test section 2 content</p>
          </TabsSection>
        </Tabs>,
      );

      const tabSection1 = container.querySelector('#tab-1');
      const tabSection2 = container.querySelector('#tab-2');

      expect(tabSection1).not.toHaveClass('govuk-tabs__panel--hidden');
      expect(tabSection2).toHaveClass('govuk-tabs__panel--hidden');

      fireEvent.click(getByText('Tab 2'));

      expect(tabSection1).toHaveClass('govuk-tabs__panel--hidden');
      expect(tabSection2).not.toHaveClass('govuk-tabs__panel--hidden');
    });
  });

  describe('single tab', () => {
    test('renders correctly', () => {
      const { container } = render(
        <Tabs>
          <TabsSection id="tab-1" title="Tab 1">
            <p>Test section 1 content</p>
          </TabsSection>
        </Tabs>,
      );

      expect(container.innerHTML).toMatchSnapshot();
    });
  });
});
