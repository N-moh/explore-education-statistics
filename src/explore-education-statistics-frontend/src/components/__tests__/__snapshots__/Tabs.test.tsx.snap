// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs multiple tabs renders correctly 1`] = `

<div class="govuk-tabs">
  <ul class="govuk-tabs__list"
      role="tablist"
  >
    <li class="govuk-tabs__list-item"
        role="presentation"
    >
      <a class="govuk-tabs__tab govuk-tabs__tab--selected"
         href="#tab-1"
         id="tab_tab-1"
         role="tab"
         aria-controls="tab-1"
         tabindex="0"
         aria-selected="true"
      >
        Tab 1
      </a>
    </li>
    <li class="govuk-tabs__list-item"
        role="presentation"
    >
      <a class="govuk-tabs__tab"
         href="#tab-2"
         id="tab_tab-2"
         role="tab"
         aria-controls="tab-2"
         tabindex="-1"
         aria-selected="false"
      >
        Tab 2
      </a>
    </li>
  </ul>
  <div class="govuk-tabs__panel"
       id="tab-1"
       role="tabpanel"
       aria-labelledby="tab_tab-1"
  >
    <p>
      Test section 1 content
    </p>
  </div>
  <div class="govuk-tabs__panel govuk-tabs__panel--hidden"
       id="tab-2"
       role="tabpanel"
       aria-labelledby="tab_tab-2"
  >
    <p>
      Test section 2 content
    </p>
  </div>
</div>

`;

exports[`Tabs renders multiple tabs correctly 1`] = `

<div class="govuk-tabs">
  <ul class="govuk-tabs__list"
      role="tablist"
  >
    <li class="govuk-tabs__list-item"
        role="presentation"
    >
      <a class="govuk-tabs__tab govuk-tabs__tab--selected"
         href="#test-section-1"
         id="tab_test-section-1"
         role="tab"
         aria-controls="test-section-1"
         tabindex="0"
         aria-selected="true"
      >
        Test section 1
      </a>
    </li>
    <li class="govuk-tabs__list-item"
        role="presentation"
    >
      <a class="govuk-tabs__tab"
         href="#test-section-2"
         id="tab_test-section-2"
         role="tab"
         aria-controls="test-section-2"
         tabindex="-1"
         aria-selected="false"
      >
        Test section 2
      </a>
    </li>
  </ul>
  <div class="govuk-tabs__panel"
       id="test-section-1"
       role="tabpanel"
       aria-labelledby="tab_test-section-1"
  >
    <p>
      Test section 1 content
    </p>
  </div>
  <div class="govuk-tabs__panel govuk-tabs__panel--hidden"
       id="test-section-2"
       role="tabpanel"
       aria-labelledby="tab_test-section-2"
  >
    <p>
      Test section 2 content
    </p>
  </div>
</div>

`;

exports[`Tabs renders single tab correctly 1`] = `

<div class="govuk-tabs">
  <ul class="govuk-tabs__list"
      role="tablist"
  >
    <li class="govuk-tabs__list-item"
        role="presentation"
    >
      <a class="govuk-tabs__tab govuk-tabs__tab--selected"
         href="#tab-1"
         id="tab_tab-1"
         role="tab"
         aria-controls="tab-1"
         tabindex="0"
         aria-selected="true"
      >
        Tab 1
      </a>
    </li>
  </ul>
  <div class="govuk-tabs__panel"
       id="tab-1"
       role="tabpanel"
       aria-labelledby="tab_tab-1"
  >
    <p>
      Test section 1 content
    </p>
  </div>
</div>

`;

exports[`Tabs single tab renders correctly 1`] = `

<div class="govuk-tabs">
  <ul class="govuk-tabs__list"
      role="tablist"
  >
    <li class="govuk-tabs__list-item"
        role="presentation"
    >
      <a class="govuk-tabs__tab govuk-tabs__tab--selected"
         href="#tab-1"
         id="tab_tab-1"
         role="tab"
         aria-controls="tab-1"
         tabindex="0"
         aria-selected="true"
      >
        Tab 1
      </a>
    </li>
  </ul>
  <div class="govuk-tabs__panel"
       id="tab-1"
       role="tabpanel"
       aria-labelledby="tab_tab-1"
  >
    <p>
      Test section 1 content
    </p>
  </div>
</div>

`;
